webauthn-security-key/libs/auth.js

```diff
@@ -0,0 +1,586 @@
+    }
+    // Set the username in the auth session so that it can be passed to the main session
+    req.session.username = username;
+    // Set the password correctness value for the next step
+    req.session.isPasswordCorrect = isPasswordCorrect(username, password);
+    // If 2fa is not set up, complete the authentication
+    const authType = getAuthType(user.credentials);
+    if (authType === SINGLE_FACTOR) {
+      completeAuthentication(req, res);
+      // If 2fa is set up, ask for a second factor
+    } else if (authType === TWO_FACTOR) {
+      res.status(200).json({
+        msg: 'Basic signin completed, 2f missing',
+        authStatus: '2fmissing',
+      });
+    } else {
+      res.status(400).json({ error: 'Unkown authentication type' });
+    }
+  }
+  // }
+);
+
+// Complete the authentication
+function completeAuthentication(req, res) {
+  // These values come from the 'auth' session
+  const { username, isPasswordCorrect } = req.session;
+  if (!username || !isPasswordCorrect) {
+    res.status(401).json({ error: 'Username or password is incorrect' });
+    return;
+  }
+  // Terminate the 'auth' session and start the 'main' session
+  // Once the 'main' session is active, the user is considered fully signed in
+  req.session.regenerate(function (err) {
+    req.session.name = 'main';
+    // Transfer the username from the old session
+    req.session.username = username;
+  });
+  req.session.save(function (err) {
+    res.status(200).json({ msg: 'Fully signed in', authStatus: 'ok' });
+  });
+}
+
+/**
+ * Respond with required information to call navigator.credential.get()
+ * Input is passed via `req.body` with similar format as output
+ * Output format:
+ * ```{
+     challenge: String,
+     userVerification: ('required'|'preferred'|'discouraged'),
+     allowCredentials: [{
+       id: String,
+       type: 'public-key',
+       transports: [('ble'|'nfc'|'usb'|'internal'), ...]
+     }, ...]
+ * }```
+ **/
+router.post('/2faOptions', csrfCheck, async (req, res) => {
+  try {
+    const user = findUserByUsername(req.session.username);
+    const userVerification = req.body.userVerification || 'required';
+    const allowCredentials = [];
+    for (let cred of user.credentials) {
+      allowCredentials.push({
+        id: cred.credId,
+        type: 'public-key',
+        transports: cred.transports,
+      });
+    }
+    const options = fido2.generateAssertionOptions({
+      timeout: FIDO_TIMEOUT,
+      rpID: process.env.HOSTNAME,
+      allowCredentials,
+      // This optional value controls whether or not the authenticator needs be able to uniquely
+      // identify the user interacting with it (via built-in PIN pad, fingerprint scanner, etc...)
+      userVerification,
+    });
+    req.session.challenge = options.challenge;
+
+    res.status(200).json(options);
+  } catch (e) {
+    res.status(400).json({
+      error: `Getting 2fa options failed: ${GENERIC_AUTH_ERROR_MESSAGE}`,
```

Comment:
@agektmr What I'm trying to reach here is that `2faOptions` and `signin2fa` generate the same error message AND that this error message does not give the attacker a hint on whether or not the password was correct. 
WDYT?